equinix/resource_metal_connection.go

```diff
@@ -340,9 +343,59 @@ func resourceMetalConnectionUpdate(d *schema.ResourceData, meta interface{}) err
 			return friendlyError(err)
 		}
 	}
+
+	// Don't update VLANs until _after_ the main ConnectionUpdateRequest has succeeded
+	if d.HasChange("vlans") {
+		connType := packngo.ConnectionType(d.Get("type").(string))
+
+		if connType == packngo.ConnectionShared {
+			old, new := d.GetChange("vlans")
+			oldVlans := convertIntArr2(old.([]interface{}))
+			newVlans := convertIntArr2(new.([]interface{}))
+			maxVlans := int(math.Max(float64(len(oldVlans)), float64(len(newVlans))))
+
+			ports := d.Get("ports").([]interface{})
+
+			for i := 0; i < maxVlans; i++ {
```

Comment:
This approach addresses most of the concerns expressed in https://github.com/equinix/terraform-provider-equinix/pull/273#discussion_r1030831300:
1. Removing a VLAN from the `vlans` attribute (i.e., changing vlans: `[a, b]` to `vlans: [a]` ) will cause `b` to be unassigned, as expected
2. Removing the first VLAN from `vlans` attribute (i.e., changing `vlans: [a, b]` to `vlans: [b]`) will cause `b` to be unassigned from the secondary port and then reassigned to the primary port, as expected
3. Swapping the VLANs (i.e., changing `vlans: [a, b]` to `vlans: [b, a]`) will cause `b` to be unassigned from the secondary port and reassigned to the primary port, but assigning `a` to the secondary port will fail with a 422 error because the circuit is still deactivating.  Re-running `terraform apply` will complete the swap.  I'm not convinced we should worry about swapping, since it's not easily supported anywhere else.